numpy/random/mtrand/mtrand.pyx

```diff
@@ -4840,12 +4841,28 @@ cdef class RandomState:
         elif isinstance(x, np.ndarray) and x.ndim > 1 and x.size:
             # Multidimensional ndarrays require a bounce buffer.
             buf = np.empty_like(x[0])
+            buffer_size = buf.nbytes
+
+            x_ptr = <char*>PyArray_DATA(<ndarray>x)
+            buf_ptr = <char*>PyArray_DATA(<ndarray>buf)
+            if PyArray_CheckExact(<ndarray>x) and PyArray_CHKFLAGS(<ndarray>x, NPY_ARRAY_CARRAY):
+                fast_buffer_copy = 1
+
             with self.lock:
                 for i in reversed(range(1, n)):
                     j = rk_interval(i, self.internal_state)
-                    buf[...] = x[j]
-                    x[j] = x[i]
-                    x[i] = buf
+                    if fast_buffer_copy:
+                        memcpy(buf_ptr, x_ptr + j * buffer_size, buffer_size)
+                        memcpy(x_ptr + j * buffer_size,
+                               x_ptr + i * buffer_size,
+                               buffer_size)
+                        memcpy(x_ptr + i * buffer_size,
+                               buf_ptr,
+                               buffer_size)
```

Comment:
Not all dtypes are `memcpy`able, as @EricCousineau-TRI will tell you

Reply:
That was a question I had too.

Reply:
Is there a reasonable method to filter these cases?

Reply:
> Not all dtypes are memcpyable, as @EricCousineau-TRI will tell you

Yuppers, captured in #10898; also just filed #11076 regarding sensitivity to memory initialization as well.

> Is there a reasonable method to filter these cases?

I think the intent post-#10898 may be to have the `REFCHK` flag more generally mean non-trivially copyable, but I myself cannot suggest anything useful for the present case...
Should we consider drafting some WIP attributes to indicate this, solidify them in future versions?